libs/ml/include/ml/dataloaders/word2vec_loaders/sgns_w2v_dataloader.hpp

```diff
@@ -0,0 +1,445 @@
+W2VLoader<T>::W2VLoader(SizeType window_size, SizeType negative_samples, bool mode)
+  : DataLoader<LabelType, DataType>(false)  // no random mode specified
+  , current_sentence_(0)
+  , current_word_(0)
+  , window_size_(window_size)
+  , negative_samples_(negative_samples)
+  , mode_(mode)
+  , target_({window_size_ * 2, 1})
+  , label_({negative_samples_, 1})
+{
+  // setup temporary buffers for training purpose
+  label_one     = LabelType({1});
+  label_one(0)  = 1;
+  label_zero    = LabelType({1});
+  label_zero(0) = 0;
+  tmp_input     = DataType({1});
+  tmp_output    = DataType({1});
+}
+
+/**
+ * reports the total size of the outputs iterating through the dataloader
+ * @tparam T
+ * @return
+ */
+template <typename T>
+math::SizeType W2VLoader<T>::Size() const
+{
+  SizeType size(0);
+  for (auto const &s : data_)
+  {
+    if ((SizeType)s.size() > (2 * window_size_))
+    {
+      size += (SizeType)s.size() - (2 * window_size_);
+    }
+  }
+
+  return size;
+}
+
+/**
+ * checks if we've passed through all the data and need to reset
+ * @tparam T
+ * @return
+ */
+template <typename T>
+bool W2VLoader<T>::IsDone() const
+{
+  if (current_sentence_ == data_.size())
+  {
+    return true;
+  }
+  else if (current_sentence_ == data_.size() - 1)  // In the last sentence
+  {
+    if (current_word_ >= data_.at(current_sentence_).size() - window_size_)
+    {
+      return true;
+    }
+  }
+  return false;
+}
+
+/**
+ * resets word cursors and re randomises negative sampling
+ * @tparam T
+ */
+template <typename T>
+void W2VLoader<T>::Reset()
+{
+  current_sentence_ = 0;
+  current_word_     = 0;
+  rng_.Seed(1337);
+  unigram_table_.Reset();
+}
+
+/**
+ * Remove words that appears less than MIN times. operation is destructive
+ * @tparam T
+ * @param min
+ */
+template <typename T>
+void W2VLoader<T>::RemoveInfrequent(SizeType min)
+{
+  W2VLoader new_loader(window_size_, negative_samples_, mode_);
+  std::map<SizeType, std::pair<std::string, SizeType>> reverse_vocab;
+  for (auto const &kvp : vocab_.data)
+  {
+    reverse_vocab[kvp.second.first] = std::make_pair(kvp.first, kvp.second.second);
+  }
+  for (auto const &sentence : data_)
+  {
+    std::string s;
+    for (auto const &word : sentence)
+    {
+      if (reverse_vocab[word].second >= min)
+      {
+        s += reverse_vocab[word].first + " ";
+      }
+    }
+    new_loader.BuildVocab(s);
+  }
+  data_       = std::move(new_loader.data_);
+  vocab_.data = std::move(new_loader.vocab_.data);
+}
+
+/**
+ * initialises the unigram table for negative frequency based sampling
+ * @tparam T
+ */
+template <typename T>
+void W2VLoader<T>::InitUnigramTable()
+{
+  std::vector<SizeType> frequencies(vocab_size());
+  for (auto const &kvp : vocab_.data)
+  {
+    frequencies[kvp.second.first] = kvp.second.second;
+  }
+  unigram_table_.Reset(1e8, frequencies);
+}
+
+template <typename T>
+void W2VLoader<T>::BufferNextSamples()
+{
+  // clear the buffer
+  buffer.clear();
```

Comment:
the problem with clearing the vector and then pushing back is that its a lot of memory allocation which is taking place inside the training loop. This will be slower than keeping a fixed array and reserving positions in that array dynamically.